iproute2: add tc-tiny reduced-size variant and reorganize dependencies

guidosarducci opened the pull request:
### Description

With the recent addition of modern BPF loader support in b048a305a3d3, `tc` gained dependencies on `libelf` and `libbpf`, with a larger installation footprint.

Following requests for a lighter-weight option (similar to `ip-tiny`), this PR splits `tc` into `tc-full` and `tc-tiny` variants, where the latter excludes the eBPF loader, uses a smaller executable, and avoids libelf and libbpf package dependencies. Both variants provide the `tc` virtual package, with `tc-tiny` as the default.

For the mips_24kc arch, the installation footprint of `tc-tiny` (without `libelf` and `libbpf`) is ~138KB smaller than `tc-full`.

The previous `tc` package included a loadable module for iptables actions. Separate this out into a common package, `tc-mod-iptables`, which both variants depend on. The old `tc` package also included a hotplug script for the link equalizer `sch_teql.ko` of package `kmod-sched`; relocate this to `kmod-sched` where it properly belongs. Both these changes were considered in the past but deferred as there was only a single `tc` package.
 
### Testing
- Build and run-tested on QEMU/malta/be32
- Verified functional differences (BPF loader) between `tc-tiny` and `tc-full`
- Verified dependencies between `tc-tiny`, `tc-full`, and `tc-mod-iptables`
- Verified working "ALTERNATIVES" support with one or both variants installed
- Confirmed hotplug script installation
- Confirmed absence of `libbpf` and `libelf` dependencies from `tc-tiny`
- Confirmed `tc-tiny` as default provider of virtual `tc` in menuconfig

### Distribution
@jow- and @ldir-EDB0 were part of past discussions on `tc` variants and the hotplug module in https://github.com/openwrt/openwrt/pull/1627.

@RussellSenior is the package maintainer, and @rsalvaterra the chief complainant urging for a size-reduced option 😄 .

@hauke reviewed/merged the previous PR https://github.com/openwrt/openwrt/pull/3904.

guidosarducci committed:
iproute2: separate tc into tiny and full variants

This change was investigated previously [1] but not deemed necessary. With
the recent addition [2] of modern BPF loader support, however, tc gained
dependencies on libelf and libbpf, with a larger installation footprint.

Similar to ip-tiny/ip-full, split tc into tc-full and tc-tiny variants,
where the latter excludes the eBPF loader, uses a smaller executable, and
avoids libelf and libbpf package dependencies. Both variants provide the
'tc' virtual package, with tc-tiny as the default.

The previous tc package included a loadable module for iptables actions.
Separate this out into a common package, tc-mod-iptables, which both
variants depend on. Some package sizes on mips_24kc:

Before:
  148343  tc_5.11.0-1_mips_24kc.ipk

After:
  144833  tc-full_5.11.0-2_mips_24kc.ipk
  138430  tc-tiny_5.11.0-2_mips_24kc.ipk  (and no libelf or libbpf)
    4115  tc-mod-iptables_5.11.0-2_mips_24kc.ipk

Also fix up some Makefile indentation.

[1] https://github.com/openwrt/openwrt/pull/1627#issuecomment-447619962
[2] b048a305a3d3 ("iproute2: update to 5.11.0")

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci committed:
kernel/modules: relocate teql hotplug from iproute2 to kmod-sched

The link equalizer sch_teql.ko of package kmod-sched relies on a hotplug
script historically included in iproute2's tc package. In previous
discussion [1], consensus was the hotplug script is best located together
with the module in kmod-sched, but this change was deferred at the time.

Relocate the hotplug script now. This change also simplifies adding a tc
variant for minimal size with reduced functionality.

[1] https://github.com/openwrt/openwrt/pull/1627#issuecomment-447923636

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

rsalvaterra reviewed (commented):
Chief complainer here, with size numbers (in bytes) for a TL-WDR3600 build (initramfs + kernel images are more interesting, since the sysupgrade ones are padded)… 😉

```
6069124 openwrt-ath79-generic-tplink_tl-wdr3600-v1-initramfs-kernel.bin (tc-full)
5957466 openwrt-ath79-generic-tplink_tl-wdr3600-v1-initramfs-kernel.bin (tc-mini)
```
… over 100 kiB saved. Nitpicks aside, this is really nice.

guidosarducci commented on package/network/utils/iproute2/Makefile:
```diff
@@ -33,29 +33,46 @@ endef
+$(call Package/iproute2/Default)
+  TITLE:=Traffic control utility (minimal)
+  VARIANT:=tctiny
+  DEFAULT_VARIANT:=1
+  PROVIDES:=tc
+  ALTERNATIVES:=200:/sbin/tc:/usr/libexec/tc-tiny
+  DEPENDS:=+kmod-sched-core +libxtables +tc-mod-iptables +(PACKAGE_devlink||PACKAGE_rdma):libmnl
 endef
 
-define Package/tc
+define Package/tc-full
 $(call Package/iproute2/Default)
-  TITLE:=Traffic control utility
-  VARIANT:=tc
+  TITLE:=Traffic control utility (full)
+  VARIANT:=tcfull
   PROVIDES:=tc
-  DEPENDS:=+kmod-sched-core +libxtables +libbpf +(PACKAGE_devlink||PACKAGE_rdma):libmnl
+  ALTERNATIVES:=300:/sbin/tc:/usr/libexec/tc-full
+  DEPENDS:=+kmod-sched-core +libxtables +tc-mod-iptables +libbpf +(PACKAGE_devlink||PACKAGE_rdma):libmnl
+endef
```
This is only loosely related to BusyBox. We use ALTERNATIVES where there's more than one exec source: `/sbin/ip` and `/sbin/tc/` have 3 and 2 sources respectively, so that's perfectly fine. Using ALTERNATIVES in this Makefile has also allowed all packages to be installed concurrently, which simplifies testing (especially for semi-maintainers like me), as well as eases switching between variants for users.

There's really no good reason I see for imposing a deliberate CONFLICTS, which would be inconsistent with the existing ALTERNATIVES usage and behaviour.

guidosarducci commented on package/network/utils/iproute2/Makefile:
```diff
@@ -33,29 +33,46 @@ endef
+  DEPENDS:=+kmod-sched-core +libxtables +tc-mod-iptables +libbpf +(PACKAGE_devlink||PACKAGE_rdma):libmnl
+endef
+
+define Package/tc-mod-iptables
+$(call Package/iproute2/Default)
+  TITLE:=Traffic control module - iptables action
+  DEPENDS:=+libxtables
 endef
```
OK, you already got it. Besides, other core packages already dynamically link with libxtables.

guidosarducci commented on package/kernel/linux/modules/netsupport.mk:
```diff
@@ -885,7 +885,7 @@ $(eval $(call KernelPackage,bpf-test))
   SUBMENU:=$(NETWORK_SUPPORT_MENU)
   TITLE:=Extra traffic schedulers
-  DEPENDS:=+kmod-sched-core +kmod-ipt-core +kmod-lib-crc32c
+  DEPENDS:=+kmod-sched-core +kmod-ipt-core +kmod-lib-crc32c +tc
```
This is just a package dependency choice, not a question of kernel/userspace layering or similar. 😃 

The key change is that now the `teql` module and `hotplug` script are bundled together in `kmod-sched`. It's almost certain that a system installing `kmod-sched` already has `tc` installed, so while it could be omitted, adding "+tc" just makes it certain and explicit.

The previous setup had the `hotplug` script pushed into `tc`, and it was very possible for someone to install `tc` without installing `kmod-sched` and the `teql` module. That's clearly a worse situation.

Both @jow- and @ldir-EDB0 were part of the prior discussion in #1627 , so I'll take their advice on whether the "+tc" dependency is wanted.

RussellSenior replied:
Another thought it to create a package that depends on the kmod and tc and that installs the script.

guidosarducci replied:
But that would split the components (script, module, tc) up into 3 different packages, which is arguably worse than what's proposed, given the current `kmod-sched` is also a monolithic, kitchen-sink package that would install lots of unneeded modules if selected by `teql-scripts`. Honestly, I would be fine with just dropping the "+tc" for now since it's unlikely to be an issue. Would that work?

Longer term, I wanted to break apart giant packages like `kmod-sched`, and e.g pull out a small `kmod-sched-teql`, which would fit nicely with your dependency idea: `tc` <-- `teql-scripts` --> `kmod-sched-teql`. But I don't think there's time enough to do all that *and* backport `iproute2` for 21.02 as I'd like.

guidosarducci replied:
After considering this for a while, I've chosen the simplest option and dropped the tc dependency to move things forward.

guidosarducci commented:
> Chief complainer here, with size numbers (in bytes) for a TL-WDR3600 build (initramfs + kernel images are more interesting, since the sysupgrade ones are padded)… 😉
> 
> ```
> 6069124 openwrt-ath79-generic-tplink_tl-wdr3600-v1-initramfs-kernel.bin (tc-full)
> 5957466 openwrt-ath79-generic-tplink_tl-wdr3600-v1-initramfs-kernel.bin (tc-mini)
> ```
> 
> … over 100 kiB saved. Nitpicks aside, this is really nice.

@rsalvaterra Great, thanks for build testing and the extra data point!

guidosarducci commented:
@RussellSenior Could you have a look at these changes too?  They add some big size reductions for `tc` in the common case and address some previously discussed and deferred cleanups (https://github.com/openwrt/openwrt/pull/1627). I've held off trying to backport the prior `iproute2` 5.11 update since these changes should also be included. Thanks for any feedback!

rsalvaterra commented:
> @rsalvaterra Great, thanks for build testing and the extra data point!

My pleasure! However, it was run-tested too, not just build-tested. 😉

guidosarducci commented:
> > @rsalvaterra Great, thanks for build testing and the extra data point!
> 
> My pleasure! However, it was run-tested too, not just build-tested. 😉

Thanks, that's even better! I just assumed as much when you only mentioned the build size numbers.

guidosarducci committed:
iproute2: add missing limits.h includes

This patch has been submitted upstream to fix an error reported by a few
users. One instance seen using gcc 10.2.0, binutils 2.35.1 and musl 1.1.24:

bpf_glue.c: In function 'get_libbpf_version':
bpf_glue.c:46:11: error: 'PATH_MAX' undeclared (first use in this function);
did you mean 'AF_MAX'?
   46 |  char buf[PATH_MAX], *s;
      |           ^~~~~~~~
      |           AF_MAX

Reported-by: Rui Salvaterra <rsalvaterra@gmail.com>
Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci commented:
Rebased against master and pushed a small update:

- dropped the tc dependency which is not strictly needed in kmod-sched
- added a patch to fix an intermittent compile error related to missing "limits.h" (also submitted upstream)

@RussellSenior I also wanted to add some context around wanting to get the iproute2 5.11 updates into OWRT 21.02. One big reason is to enable a key BPF feature available with LTS kernel 5.4, namely support for "C" programs using initialized global variables. By switching to use `libbpf`, iproute2 5.11 is the first version to support loading these kinds of BPF programs, and "tc-tiny" from this PR eliminates the `libbpf` size overhead for non-BPF users.

hauke commented:
After these changes we have 4 build variants, it is possible to reduce this to only one tiny and one full variant and take the ip-tiny  and tc-tiny from the tiny build variant?
It would also be nice to remove the dependency to ```+(PACKAGE_devlink||PACKAGE_rdma):libmnl``` in the tiny version.

guidosarducci commented:
> After these changes we have 4 build variants, it is possible to reduce this to only one tiny and one full variant and take the ip-tiny and tc-tiny from the tiny build variant?

I explored that as the more preferable option but it's not possible. Doing so would require both "tiny" utilities to share one set of config options, while the "full" utilities share another set of config options. However, `tc-tiny` and `ip-tiny` themselves need different settings for `SHARED_LIBS`, as do `tc-full` and `ip-full`, so there's no symmetry there.

> It would also be nice to remove the dependency to `+(PACKAGE_devlink||PACKAGE_rdma):libmnl` in the tiny version.

I remember looking into this a while ago and it's not easy/possible. All the utilities have a `libmnl` dependency because they link with a common internal library (libnetlink.a) that pulls in `libmnl`. However, it's not really that bad: at 6.5KB on mips_24kc, the `libmnl` package is tiny compared to `libelf` and `libbpf`, and also has other users besides iproute2:
```
# opkg whatdepends libmnl0
Root set:
  libmnl0
What depends on root set
        ip-full 5.0.0-2.1       depends on libmnl0
        libnetfilter-conntrack3 2018-05-01-3ccae9f5-2   depends on libmnl0
        tc 5.0.0-2.1    depends on libmnl0
        ipset 7.3-1     depends on libmnl0
        wireguard-tools 1.0.20191226-1  depends on libmnl0
        libipset13 7.3-1        depends on libmnl0
        dnsmasq-full 2.80-16.3  depends on libnetfilter-conntrack3
```
I don't think there's much more to squeeze out in terms of size savings. Together, https://github.com/openwrt/openwrt/commit/b048a305a3d3cbb5375e32757e70aa129b4c7c4f and this PR already dropped some dependencies and reduced footprint for the baseline installation, while offering feature improvements and fixes wanted for a long time. 

Can we consider merging this now? I assume you're still OK with these changes @RussellSenior ?

guidosarducci commented:
The patch a7c57d455ee3  ("iproute2: add missing limits.h includes") has now been [accepted upstream](https://patchwork.kernel.org/project/netdevbpf/patch/20210311214754.3566712-1-Tony.Ambardar@gmail.com/).

hauke commented:
Thank you for the pull request, I pushed it to master.

guidosarducci commented:
Thanks, Hauke.
